課題１

takurooo opened the pull request:
# 概要
静止画像のフォーマット変換を行うツール  
指定されたディレクトリ下にある静止画像を全てフォーマット変換する。  
オプションで変換前のフォーマットと変換後のフォーマットを指定することができる。

# 対応フォーマット
- jpeg
- png


# ビルド
``go build -o imgtrans cmd/imgtrans/imgtrans.go``

# 使い方

### Usage
```
Usage of imgtrans:
  -d string
    	directory path
  -i string
    	input format(jpg or png) (default "jpg")
  -o string
    	output formatjpg or png (default "png")
```

### Examples
例1)  
testディレクトリ以下にある全てのjpg or jpegファイルをpngフォーマットに変換  
``imgtrans -d ./test``  

例2)  
testディレクトリ以下にある全てのpngファイルをjpgフォーマットに変換  
``imgtrans -i png -o jpg -d ./test``

takurooo committed:
add new files

takurooo committed:
delete files

takurooo committed:
add comment

takurooo committed:
wrap error

takurooo committed:
add README

takurooo committed:
add CanTrans function

ebiiim commented on kadai1/takurooo/transcoder/transcoder.go:
```diff
@@ -0,0 +1,122 @@
+	switch format {
+	case PNG:
+		img, err = png.Decode(r)
+	case JPEG:
+		fallthrough
+	case JPG:
+		img, err = jpeg.Decode(r)
```
```suggestion
	case JPEG, JPG:
		img, err = jpeg.Decode(r)
```

takurooo replied:
caseに複数の値書けるんですね!!修正しました。

ebiiim commented on kadai1/takurooo/transcoder/transcoder.go:
```diff
@@ -0,0 +1,122 @@
+	if err != nil {
+		return err
+	}
+	defer rf.Close()
```
エラーを確認したほうがいいです

takurooo replied:
defer xxx.Close() している箇所はエラー確認するようにしました。

ebiiim commented on kadai1/takurooo/cmd/imgtrans/imgtrans.go:
```diff
@@ -0,0 +1,97 @@
+		flag.Usage()
+		return ExitError
+	}
+	{
```
ここはブロックに入れなくていい気がします・・・？

takurooo replied:
削除します...(エラーチェックの変数とそのあとに出てくる変数名がかぶるのが気持ち悪かったのですが、不要ですね...)

ebiiim commented on kadai1/takurooo/transcoder/transcoder.go:
```diff
@@ -0,0 +1,122 @@
+	// encode phase
+	// -------------------------
+	dir := filepath.Dir(inPath)
+	base := filepath.Base(inPath[:len(inPath)-len(filepath.Ext(inPath))])
```
```suggestion
	base := filepath.Base(strings.TrimRight(inPath, filepath.Ext(inPath)))
```
ライブラリを使って書くこともできます。読みやすさはなんとも言えないですが...

takurooo replied:
strings.TrimRightだとjpg.jpgとかでうまくいかなそうだったので、strings.TrimSuffix置き換えました！

ebiiim commented on kadai1/takurooo/transcoder/transcoder.go:
```diff
@@ -0,0 +1,122 @@
+		return strings.Contains(ext, JPG) || strings.Contains(ext, JPEG)
+	}
+
+	return strings.Contains(ext, t.inFormat)
```
`hoge.jjjjjjjjpeggggggg`が通る気がします

takurooo replied:
ありがとうございます。バグですね....
extのドットを削って単純な文字列比較するように修正しました！

takurooo committed:
- transcoder.goのcase文修正
- Close()のエラーチェック追加
- 不要なブロック削除
- strings.TrimSuffixを使用するように修正
- CanTrans()のチェック方法を修正
